src/main/scala/tech/sourced/gemini/ConnectedComponents.scala

```diff
@@ -0,0 +1,125 @@
+package tech.sourced.gemini
+
+import java.nio.ByteBuffer
+
+import com.datastax.driver.core.{Session, SimpleStatement}
+import org.slf4j.{Logger => Slf4jLogger}
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+
+case class FileHash(sha1: String, value: ByteBuffer)
+
+class DBConnectedComponents(
+                              log: Slf4jLogger,
+                              conn: Session,
+                              keyspace: String = Gemini.defautKeyspace) extends ConnectedComponents(log) {
+  def getHashtables(): List[Byte] = {
+    val hashtablesCql = s"SELECT DISTINCT hashtable FROM $keyspace.hashtables"
+
+    conn
+      .execute(new SimpleStatement(hashtablesCql))
+      .asScala
+      .map(_.getByte("hashtable"))
+      .toList.
+      sorted
+  }
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash] = {
+    val cql = s"SELECT sha1, value FROM $keyspace.hashtables WHERE hashtable=$hashtable"
+
+    conn.execute(new SimpleStatement(cql))
+      .asScala
+      .map(row => FileHash(row.getString("sha1"), row.getBytes("value")))
+  }
+}
+
+abstract class ConnectedComponents(log: Slf4jLogger) {
+  def getHashtables(): List[Byte]
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash]
+
+  def makeBuckets(): List[List[Int]] = {
+    val (buckets, elementIds) = getHashtables()
+      .foldLeft(List[List[Int]](), collection.mutable.Map[String, Int]()) {
+        (result, hashtable) =>
+
+          var (buckets, elementIds) = result
+          val prevBucketsSize = buckets.size
+          var band: Option[ByteBuffer] = None
+          var bucket = List[Int]()
+
+          getHashValues(hashtable).foreach { case FileHash(sha1, value) =>
+            val elId = elementIds.getOrElseUpdate(sha1, elementIds.size)
+            if (!band.contains(value)) {
+              if (band.isDefined) {
+                buckets = buckets :+ bucket
+                bucket = List[Int]()
+              }
+              band = Some(value)
+            }
+            bucket = bucket :+ elId
+          }
+
+          if (bucket.nonEmpty) {
+            buckets = buckets :+ bucket
+          }
+
+          val bucketsSize = buckets.size - prevBucketsSize
+          log.info(s"Fetched $hashtable, $bucketsSize buckets")
+
+          (buckets, elementIds)
+    }
+
+    log.info(s"Number of buckets: ${buckets.size}")
+    log.info(s"Number of elements: ${elementIds.size}")
+
+    buckets
+  }
+
+  def elementsToBuckets(buckets: List[List[Int]]): Map[Int, List[Int]] = {
+    buckets.zipWithIndex.foldLeft(Map[Int, List[Int]]()) { case (elementToBuckets, (bucket, i)) =>
+      bucket.foldLeft(elementToBuckets) { (elementToBuckets, el) =>
+        var list = elementToBuckets.getOrElse(el, List[Int]())
+        list = list :+ i
+        elementToBuckets + (el -> list)
+      }
+    }
+  }
+
+  def findInBuckets(
```

Comment:
It may be a good idea to add some [ScalaDoc](http://twitter.github.io/effectivescala/#Formatting-Comments) with very high-level description of the steps of the algorithm, so the reader has some idea of what is being done below.